### ISSUE TITLE
Inconsistence between train mode and fork mode

### ISSUE THREAD

**Issue by juunnn** [facebookresearch/wav2letter#647]
Hi, I am trying to make training with two dataset, one with 300 h of data and other with 30 h data. One I training with 300 h of data and got pretty good WER, so I attempt to fork this model and training it with my other data. But when I fork the model with 30 h dataset, it produces NaN Loss value.
```sh
F0508 11:37:30.240669 82526 Train.cpp:564] Loss has NaN values. Samples - fa520080-13d9-4aca-9352-1bcd6d02460a,c6be0956-2171-43c5-ac60-8f8612eafe47
```
Both dataset have the same format, all the data in 300 h and 30 h have sample rate in 16000, 16 bit, have the same .wav format I can be assured that the data have the same format what so ever.

I am curious whether I have something wrong with that dataset, so I tried to train with this command
```sh
mpirun -n 4 --allow-run-as-root $w2l/Train train --flagfile config/stream/forkMK.cfg 
```
the process runs fine. Got bad WER but not have NaN values on the loss (tried this with 20 iters).
I use the same config files and the same arch file for both fork and train. Do I have something wrong with the fork concept? because I read at the wiki, fork mode can be used to adapt a model with a new dataset. 
Thank you for your answer.

**Comment by lunixbochs** [comment]
which model architecture and spoken language?

**Comment by juunnn** [comment]
[Convnet](https://github.com/facebookresearch/wav2letter/tree/master/recipes/models/streaming_convnets/librispeech) model and my own dataset (Bahasa Indonesia).

**Comment by padentomasello** [comment]
hi @juunnn,  sorry for the trouble. Would you mind providing the exact commands you are using to reproduce, and the logs of the output? 

Some other ideas that might be helpful to debug in the meantime:

1. As a sanity check, if you run `fork` on the same 300h data, do the result look as expected?

2. Does the loss immediately go to NAN? Or does it slowly diverge? 

3. If you run `fork` on the 30h with zero learning rate, do you still get NAN values? (Sometimes when fine-tuning models, you need to be careful about setting the learning rate. Particularly if the pretraining model finished training with a lower learning rate than you're currently using).

**Comment by juunnn** [comment]
Thank you for your suggestion @padentomasello, I have tried it and got the result.

1. I tried fork with the 300 h data, it works fine. The learning rate is set to 0.1 and the loss (event still high) not NaN. It starts high but slowly degrades as iteration increase. This is the log from what I got
```sh
I0513 19:42:56.663806  2454 Train.cpp:328] epoch:        1 | lr: 0.100000 | lrcriterion: 0.000000 | runtime: 02:04:10 | bch(ms): 307.96 | smp(ms): 37.86 | fwd(ms): 86.33 | crit-fwd(ms): 2.77 | bwd(ms): 165.80 | optim(ms): 17.71 | loss:    5.13663 | train-TER: 13.67 | train-WER: 36.44 | test-loss:    3.66644 | test-TER:  6.93 | test-WER: 21.30 | avg-isz: 694 | avg-tsz: 089 | max-tsz: 208 | hrs:  373.63 | thrpt(sec/sec): 180.52
I0513 19:42:57.624686  2454 Train.cpp:531] Shuffling trainset
I0513 19:42:57.628823  2454 Train.cpp:538] Epoch 2 started!
I0513 21:45:28.341117  2454 Train.cpp:328] epoch:        2 | lr: 0.100000 | lrcriterion: 0.000000 | runtime: 01:58:07 | bch(ms): 292.92 | smp(ms): 0.56 | fwd(ms): 96.32 | crit-fwd(ms): 3.11 | bwd(ms): 177.85 | optim(ms): 17.98 | loss:    4.72057 | train-TER: 13.66 | train-WER: 36.04 | test-loss:    3.58320 | test-TER:  6.79 | test-WER: 20.71 | avg-isz: 694 | avg-tsz: 089 | max-tsz: 208 | hrs:  373.63 | thrpt(sec/sec): 189.79
I0513 21:45:30.927461  2454 Train.cpp:531] Shuffling trainset
I0513 21:45:30.932426  2454 Train.cpp:538] Epoch 3 started!
I0513 23:49:00.487404  2454 Train.cpp:328] epoch:        3 | lr: 0.100000 | lrcriterion: 0.000000 | runtime: 01:59:03 | bch(ms): 295.26 | smp(ms): 0.56 | fwd(ms): 97.10 | crit-fwd(ms): 3.15 | bwd(ms): 179.29 | optim(ms): 18.09 | loss:    4.52207 | train-TER: 13.51 | train-WER: 35.76 | test-loss:    3.65223 | test-TER:  6.44 | test-WER: 19.58 | avg-isz: 694 | avg-tsz: 089 | max-tsz: 208 | hrs:  373.63 | thrpt(sec/sec): 188.29
I0513 23:49:03.143146  2454 Train.cpp:531] Shuffling trainset
I0513 23:49:03.147814  2454 Train.cpp:538] Epoch 4 started!
I0514 01:53:29.154036  2454 Train.cpp:328] epoch:        4 | lr: 0.100000 | lrcriterion: 0.000000 | runtime: 01:59:54 | bch(ms): 297.39 | smp(ms): 0.60 | fwd(ms): 97.72 | crit-fwd(ms): 3.18 | bwd(ms): 180.67 | optim(ms): 18.19 | loss:    4.39018 | train-TER: 12.82 | train-WER: 33.89 | test-loss:    3.89929 | test-TER:  6.32 | test-WER: 19.36 | avg-isz: 694 | avg-tsz: 089 | max-tsz: 208 | hrs:  373.63 | thrpt(sec/sec): 186.94
I0514 01:53:31.888656  2454 Train.cpp:531] Shuffling trainset
I0514 01:53:31.894687  2454 Train.cpp:538] Epoch 5 started!
I0514 03:58:23.200577  2454 Train.cpp:328] epoch:        5 | lr: 0.100000 | lrcriterion: 0.000000 | runtime: 02:00:23 | bch(ms): 298.59 | smp(ms): 0.59 | fwd(ms): 98.11 | crit-fwd(ms): 3.17 | bwd(ms): 181.47 | optim(ms): 18.20 | loss:    4.32636 | train-TER: 12.69 | train-WER: 33.50 | test-loss:    3.84318 | test-TER:  6.44 | test-WER: 19.66 | avg-isz: 694 | avg-tsz: 089 | max-tsz: 208 | hrs:  373.63 | thrpt(sec/sec): 186.19
I0514 03:58:24.624155  2454 Train.cpp:531] Shuffling trainset
```
3. When I use 0.0 learning rate to fork the 30 h dataset, it was immediately return NaN values (without even 1st epoch successfully)
```sh
I0514 20:46:41.649883  2565 Train.cpp:538] Epoch 1 started!
F0514 20:46:56.396183  2565 Train.cpp:564] Loss has NaN values. Samples - 8451a50f-a4b6-4210-aeb2-73ddc2959e9f
*** Check failure stack trace: ***
    @     0x7fc5625435cd  google::LogMessage::Fail()
    @     0x7fc562545433  google::LogMessage::SendToLog()
    @     0x7fc56254315b  google::LogMessage::Flush()
    @     0x7fc562545e1e  google::LogMessageFatal::~LogMessageFatal()
    @           0x4942d8  _ZZ4mainENKUlSt10shared_ptrIN2fl6ModuleEES_IN3w2l17SequenceCriterionEES_INS3_10W2lDatasetEES_INS0_19FirstOrderOptimizerEES9_ddbiE3_clES2_S5_S7_S9_S9_ddbi.constprop.12666
    @           0x41c280  main
    @     0x7fc5618ed830  __libc_start_main
    @           0x48e289  _start
    @              (nil)  (unknown)

```

The command I used for both training is the same, the different is only on the train and valid tag.
This is the snipset of the flag file,
```sh
--runname=streamingMK
--rundir=/data/jun/training/exp

--tokensdir=/data/jun/training/data/MK/txt #This is the different part
--archdir=config/stream

--datadir=/data/jun/training/data/MK/lists #This is the different part
--train=train
--valid=test

--lexicon=/data/jun/training/data/MK/txt/lexicon #This part also different
--tokens=tokens
--arch=am_500ms_future_context.arch

--criterion=ctc
--batchsize=1
--lr=0.0
--lrcosine=true
--momentum=0.0
--maxgradnorm=0.5

#--critoptim=adam
#--netoptim=adam
#--reportiters=1000
--iter=10

--nthread=1
--input=wav
--mfsc=true
--usewordpiece=true
--wordseparator=|
--filterbanks=80
--minisz=200
--mintsz=2
--maxisz=33000

--enable_distributed=true
--pcttraineval=1
--minloglevel=0
--logtostderr
--onorm=target
--sqnorm
--localnrmlleftctx=300

--alsologtostderr
```
and this is the command I used for both experiment
```sh
/root/wav2letter/build/Train fork exp/streaming/002_model_last.bin --flagfile fork.cfg
```
And also I tried using the newest docker image that I found on the docker hub, but it also not working. Thank you for your help.

**Comment by padentomasello** [comment]
@juunnn Thanks for the detailed response. 

> I tried fork with the 300 h data, it works fine. The learning rate is set to 0.1 and the loss (event still high) not NaN. It starts high but slowly degrades as iteration increase. This is the log from what I got

What was the ending WER for the initial model? Also, what was the lr on the last epoch of the baseline model?

Are the tokens different for the different datasets? If possible, it could help to use the same tokens for both datasets. 

Also, I noticed you had `adam` commented out in your config. Was this use for the initial training job? If so, I would try retraining the baseline with SGD. Adam has some internal state which may make forking from a model trained with it numerically unstable. 

I should mention, fine tuning is not an exact science, so while I'm offering suggestions of things to look into, there is no guarantee that a model can be fine tuned on other datasets and this is an active area of research.

**Comment by padentomasello** [comment]
@juunnn  You may also be interested in https://github.com/facebookresearch/wav2letter/issues/507#issuecomment-617530809 if you haven't seen it already.

**Comment by juunnn** [comment]
> Are the tokens different for the different datasets? If possible, it could help to use the same tokens for both datasets.

I double-check the tokens file, it sure has different numbers of items. I don't expect this, because the data preparation never has a problem before. Thank you for your answer.

**Comment by padentomasello** [comment]
@juunnn Interesting. It may be worth trying to use the same tokens for both datasets. (Assuming it makes sense to do so?). 

Some other thoughts:

Are the datasets from similar sources? If the datasets are different, (one from a audio book, another from tv), fine tuning is going to be very challenging. 

If you create a a baseline training job on 300h, but run it for 0 epochs, (essentially just initialize the model, and dump it), do you still get Nans when trying to fork on 30h? If so, this may indicate a bug in the config or implementation, and not an ML "bug". 

Are you able to train a model from scratch using 30h? 

Does the baseline overfit to the training data? 300 hours is not too large, where we might begin to overfit on the training data, and consequently the model will not generalize well or having NAN values when trying to fine tune. 

Otherwise, like I mentioned before, this is not an exact science. Fine tuning models can be challenging, and I'd encourage you to investigate other techniques.

**Comment by juunnn** [comment]
> Are the datasets from similar sources? If the datasets are different, (one from a audio book, another from tv), fine-tuning is going to be very challenging.
yeah, they are from a different source. the baseline from reading (which have great quality) and the 30h data is from free speech (monologue without reference). I got like 64% WER from that while the baseline only produces like 9% WER.

After changing the tokens I think the forking working good. But I have a question, what is the mutable flags when we fork from another model? I still don't know what can and can not we change when we do forking and continue mode.

**Comment by tlikhomanenko** [comment]
In case of continue you will reuse optimizer - so even if you change it it will be loaded from the model. In case of fork new optimizer will be created with respect to flags params. Also for both continue and fork modes network and criterions will be reused from the model and all params on that will be ignored. The rest flags will be used directly from updated version or from saved state stored in the model if you son't provide them in cmd line.
